sdk/src/program_error.rs

```diff
@@ -0,0 +1,129 @@
+    NotEnoughAccountKeys,
+    /// Failed to borrow a reference to an account, already borrowed
+    AccountBorrowFailed,
+}
+
+/// 32bit representations of builtin program errors returned by the entry point
+const BUILTIN_ERROR_START: u32 = 0x8000_0000; // 31st bit set
+const INVALID_ARGUMENT: u32 = BUILTIN_ERROR_START + 1;
+const INVALID_INSTRUCTION_DATA: u32 = BUILTIN_ERROR_START + 2;
+const INVALID_ACCOUNT_DATA: u32 = BUILTIN_ERROR_START + 3;
+const ACCOUNT_DATA_TOO_SMALL: u32 = BUILTIN_ERROR_START + 4;
+const INSUFFICIENT_FUNDS: u32 = BUILTIN_ERROR_START + 5;
+const INCORRECT_PROGRAM_ID: u32 = BUILTIN_ERROR_START + 6;
+const MISSING_REQUIRED_SIGNATURES: u32 = BUILTIN_ERROR_START + 7;
+const ACCOUNT_ALREADY_INITIALIZED: u32 = BUILTIN_ERROR_START + 8;
+const UNINITIALIZED_ACCOUNT: u32 = BUILTIN_ERROR_START + 9;
+const NOT_ENOUGH_ACCOUNT_KEYS: u32 = BUILTIN_ERROR_START + 10;
+const ACCOUNT_BORROW_FAILED: u32 = BUILTIN_ERROR_START + 11;
+
+/// Is this a builtin error? (is 31th bit set?)
+fn is_builtin(error: u32) -> bool {
+    (error & BUILTIN_ERROR_START) != 0
+}
+
+/// If a program defined error conflicts with a builtin error
+/// its 30th bit is set before returning to distinguish it.
+/// The side effect is that the original error's 30th bit
+/// value is lost, be aware.
+const CONFLICTING_ERROR_MARK: u32 = 0x4000_0000; // 30st bit set
+
+/// Is this error marked as conflicting? (is 30th bit set?)
+fn is_marked_conflicting(error: u32) -> bool {
+    (error & CONFLICTING_ERROR_MARK) != 0
+}
+
+/// Mark as a conflicting error
+fn mark_conflicting(error: u32) -> u32 {
+    error | CONFLICTING_ERROR_MARK
+}
+
+/// Unmark as a conflicting error
+fn unmark_conflicting(error: u32) -> u32 {
+    error & !CONFLICTING_ERROR_MARK
+}
+
+impl From<ProgramError> for u32 {
+    fn from(error: ProgramError) -> Self {
+        match error {
+            ProgramError::InvalidArgument => INVALID_ARGUMENT,
+            ProgramError::InvalidInstructionData => INVALID_INSTRUCTION_DATA,
+            ProgramError::InvalidAccountData => INVALID_ACCOUNT_DATA,
+            ProgramError::AccountDataTooSmall => ACCOUNT_DATA_TOO_SMALL,
+            ProgramError::InsufficientFunds => INSUFFICIENT_FUNDS,
+            ProgramError::IncorrectProgramId => INCORRECT_PROGRAM_ID,
+            ProgramError::MissingRequiredSignature => MISSING_REQUIRED_SIGNATURES,
+            ProgramError::AccountAlreadyInitialized => ACCOUNT_ALREADY_INITIALIZED,
+            ProgramError::UninitializedAccount => UNINITIALIZED_ACCOUNT,
+            ProgramError::NotEnoughAccountKeys => NOT_ENOUGH_ACCOUNT_KEYS,
+            ProgramError::AccountBorrowFailed => ACCOUNT_BORROW_FAILED,
+            ProgramError::CustomError(error) => {
+                if error == 0 || is_builtin(error) {
+                    mark_conflicting(error)
+                } else {
+                    error
+                }
+            }
+        }
+    }
+}
+
+impl<T> From<T> for InstructionError
```

Comment:
Thoughts on this living in `sdk/instruction.rs`?

Reply:
It was in `instruction_procesor_utils.rs` and being it's so linked to program error (pretty much the only folks doing `u32` conversion to `InstructionError` would be programs I'm thinking `program_error.rs` is the best choice.